sql/catalyst/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala

```diff
@@ -55,7 +56,52 @@ case class DataSourceV2Relation(
   }
 
   override def computeStats(): Statistics = {
-    val scan = newScanBuilder().build()
+    if (Utils.isTesting) {
+      // when testing, throw an exception if this computeStats method is called because stats should
+      // not be accessed before pushing the projection and filters to create a scan. otherwise, the
+      // stats are not accurate because they are based on a full table scan of all columns.
+      throw new UnsupportedOperationException(
+        s"BUG: computeStats called before pushdown on DSv2 relation: $name")
+    } else {
+      // when not testing, return stats because bad stats are better than failing a query
+      newScanBuilder() match {
+        case r: SupportsReportStatistics =>
+          val statistics = r.estimateStatistics()
+          DataSourceV2Relation.transformV2Stats(statistics, None, conf.defaultSizeInBytes)
+        case _ =>
+          Statistics(sizeInBytes = conf.defaultSizeInBytes)
+      }
+    }
+  }
+
+  override def newInstance(): DataSourceV2Relation = {
+    copy(output = output.map(_.newInstance()))
+  }
+}
+
+/**
+ * A logical plan for a DSv2 table with a scan already created.
+ *
+ * This is used in the optimizer to push filters and projection down before conversion to physical
+ * plan. This ensures that the stats that are used by the optimizer account for the filters and
+ * projection that will be pushed down.
+ *
+ * @param table a DSv2 [[Table]]
+ * @param scan a DSv2 [[Scan]]
+ * @param output the output attributes of this relation
+ */
+case class DataSourceV2ScanRelation(
```

Comment:
It seems to me that the maintenance cost is high if we have 2 nodes for ds v2 relation. Whenever we need to pattern match the v2 relation node, we need to take care of the 2 nodes both.

How about
```
case class DataSourceV2Relation(
    table: Table,
    output: Seq[AttributeReference],
    options: CaseInsensitiveStringMap,
    createdScan: Option[Scan])

def getScan() = {
  if (isTest) {
    assert(createdScan.isDefined)
    createdScan.get
  } else {
    createdScan.getOrElse ...
  }
}
```

Reply:
This should be a separate node to avoid correctness problems. Otherwise, it is easy to accidentally write rules that match both `DataSourceV2Relation` and `DataSourceV2ScanRelation` but does not handle the case where operators have already been pushed.

When a filter is pushed down, it is also removed from the filters on top of the scan. If push-down happens a second time because rules match the same node, then it is easy for a mistake to ignore the original pushed filter and create a second independent scan. That's a correctness bug that is easy to introduce by accident.

Using a separate relation type requires rules to choose whether to support a relation after push-down, or just a relation before push-down. The trade-off is that some places need to match both, but those cases are few and worth the trade.

Reply:
Makes sense. But it's still better to reduce the maintenance cost a bit. How about we clearly define the life cycle of a v2 relation as: `UnresolvedRelation` (before analysis) -> `DataSourceV2Relation` (after analysis) -> `DataSourceV2ScanRelation` (after optimization) -> the physical scan node.

Then we only need to match one node:
1. if dealing with `analyzedPlan`, match `DataSourceV2Relation`
2. if dealing with `optimizedPlan`, match `DataSourceV2ScanRelation`

Reply:
I did this to simplify the code. Now places that use `optimizedPlan` expect the scan relation.

There was one minor problem with this. To address the concern about relations used by DDL commands like `AlterTable` getting modified, those plans no longer list the relation as a child, so rules are not automatically run on it. For those DDL commands, `DataSourceV2Relation` is still in the optimized plan because it is no longer converted. I think this is correct behavior.

Reply:
I think this is fine. DDL commands don't list the relation as a child, and only the corresponding planner rule need to match the `DataSourceV2Relation`.